packages/contracts-bedrock/scripts/DeploySuperchain.s.sol

```diff
@@ -0,0 +1,384 @@
+    // inputs reverts if this flag is false. This ensures the deploy script cannot proceed until all
+    // inputs are validated and set.
+    bool public inputSet = false;
+
+    // The full input struct is kept in storage. It is not exposed because the return type would be
+    // a tuple, but it's more convenient for the return type to be the struct itself. Therefore the
+    // struct is exposed via the `input()` getter method.
+    Input internal inputs;
+
+    // And each field is exposed via it's own getter method. We can equivalently remove these
+    // storage variables and add getter methods that return the input struct fields directly, but
+    // that is more verbose with more boilerplate, especially for larger scripts with many inputs.
+    // Unlike the `input()` getter, these getters do not revert if the input is not set. The caller
+    // should check the `inputSet` value before calling any of these getters.
+    address public proxyAdminOwner;
+    address public protocolVersionsOwner;
+    address public guardian;
+    bool public paused;
+    ProtocolVersion public requiredProtocolVersion;
+    ProtocolVersion public recommendedProtocolVersion;
+
+    // Load the input from a TOML file.
+    function loadInputFile(string memory _infile) public {
+        _infile;
+        Input memory parsedInput;
+        loadInput(parsedInput);
+        require(false, "DeploySuperchainInput: loadInput is not implemented");
+    }
+
+    // Load the input from a struct.
+    function loadInput(Input memory _input) public {
+        // As a defensive measure, we only allow inputs to be set once.
+        require(!inputSet, "DeploySuperchainInput: Input already set");
+
+        // All assertions on inputs happen here. You cannot set any inputs in Solidity unless
+        // they're all valid. For Go testing, the input and outputs
+        require(_input.roles.proxyAdminOwner != address(0), "DeploySuperchainInput: Null proxyAdminOwner");
+        require(_input.roles.protocolVersionsOwner != address(0), "DeploySuperchainInput: Null protocolVersionsOwner");
+        require(_input.roles.guardian != address(0), "DeploySuperchainInput: Null guardian");
+
+        // We now set all values in storage.
+        inputSet = true;
+        inputs = _input;
+
+        proxyAdminOwner = _input.roles.proxyAdminOwner;
+        protocolVersionsOwner = _input.roles.protocolVersionsOwner;
+        guardian = _input.roles.guardian;
+        paused = _input.paused;
+        requiredProtocolVersion = _input.requiredProtocolVersion;
+        recommendedProtocolVersion = _input.recommendedProtocolVersion;
+    }
+
+    function input() public view returns (Input memory) {
+        require(inputSet, "DeploySuperchainInput: Input not set");
+        return inputs;
+    }
+}
+
+contract DeploySuperchainOutput {
+    // The output struct contains all the output data from the deployment.
+    struct Output {
```

Comment:
The current solution offers both typed contract interfaces (good for Go) + structs (meant for tests), but I still think the former has some benefits for tests. Here's my thinking:

If we expand the input or output scope, e.g. in this case by adding more implementation-contracts, that OP-chains can hook their proxies onto, then this `struct` changes.

If we return the `DeploySuperchainOutput` contract address, instead of the struct, then a test can fetch the attributes it needs, and ignore the other ones. E.g. it could read a struct for a specialized output subset of the deployment, or just an attribute at a time.

And potentially that means the test bytecode doesn't change (no new different struct to unpack), when you add an attribute that the test doesn't need, as the interface expands but the handling doesn't change.

So if a deploy-contract gets reused between a lot of tests, and you modify the deploy contract, then all the tests don't change their bytecode. Perhaps with interfaces it's even possible to avoid entirely recompiling the tests, by forcing a specific smaller interface.

The impact is small now, but that can change once we add stuff to contracts, have larger deploy contracts (like the Op chain deployment) with more frequent changes, and more consumers (tests, other scripts). At that point I think compile times can (at least theoretically) improve from not using the struct pattern (assuming the deploy input / output is large).

-----

On the other hand though, the struct UX is nice for initializing an input state in a test. And that does end up requiring completeness + recompiling. If only there was some way to solve that, and then deployers could be larger without negative impact on tests / compiling.
For now, keeping the new deployer inputs/outputs small is probably best anyway.

Reply:
These are good points. I think the value of type contract interfaces for solidity tests will become more apparent once we add support deployment of features like Alt-DA and custom gas token through these scripts. I think that will likely be done with an inheritance-based approach similar to https://github.com/ethereum-optimism/design-docs/blob/main/smart-contract-feature-development.md, to have a different deploy script that inherits the base one.

In 9cf7aec4a I did add a getter method to return the dsi/dso addresses, which will make migrating to that pattern simpler once it's needed